### PR TITLE
[qradar] fix mirroring issue

### PR DESCRIPTION
**Pull request by ilaner**
This will save two update times in the context: one for modified offenses and one for closed offenses

fixes: https://jira-dc.paloaltonetworks.com/browse/CIAC-8821

### CODE REVIEW

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py [hunk 1]
```diff
@@ -3698,17 +3705,23 @@ def get_modified_remote_data_command(client: Client, params: dict[str, str],
...
-    last_update_time = ctx.get(LAST_MIRROR_KEY, 0)
-    if not last_update_time:
-        last_update_time = remote_args.last_update
-    last_update = get_time_parameter(last_update_time, epoch_format=True)
-    assert isinstance(last_update, int)
-    filter_modified = f'id <= {highest_fetched_id} AND status!=closed AND last_persisted_time > {last_update}'
-    filter_closed = f'id <= {highest_fetched_id} AND status=closed AND close_time > {last_update}'
+    last_update_modified = ctx.get(LAST_MIRROR_KEY, 0)
+    if not last_update_modified:
+        # it means wer are in the first mirror. We get the last update of the latest incident with a window of 5 minutes
```
Comment (GuyAfik) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:3715]:
```suggestion
        # This is the first mirror. We get the last update of the latest incident with a window of 5 minutes
```

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py [hunk 2]
```diff
@@ -3639,7 +3646,7 @@ def add_modified_remote_offenses(client: Client,
         new_context_data.update({MIRRORED_OFFENSES_QUERIED_CTX_KEY: mirrored_offenses_queries})
         new_context_data.update({MIRRORED_OFFENSES_FINISHED_CTX_KEY: finished_offenses_queue})
 
-    new_context_data.update({LAST_MIRROR_KEY: current_last_update})
+    new_context_data.update({LAST_MIRROR_KEY: new_last_update_modified, LAST_MIRROR_CLOSED_KEY: new_last_update_closed})
```
Comment (GuyAfik) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:3649]: didn't we say we don't need additional key in context for closed incidents?

File: Packs/QRadar/ReleaseNotes/2_4_40.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+
+#### Integrations
+
+##### IBM QRadar v3
+
+- Fixed an issue with mirroring QRadar offenses where some offenses were failed to mirror.
```
Comment (GuyAfik) [inline Packs/QRadar/ReleaseNotes/2_4_40.md:6]:
```suggestion
- Fixed an issue with mirroring where closing offenses in Qradar didn't close the equivalent xsoar incident in some cases.  
```

**Review by GuyAfik — APPROVED** [review]

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py (hunk 2 above)
Comment (ilaner) [reply Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:3649]: There is another edge case:
1. We query for modified and closed offenses.
2. The latest modified time is 3. The latest closed time is 7.
3. We will save 7 as the last update time.
4. In the next mirroring run, we can miss offenses that were modified between 4-6.

The safer option is just to store two separate timestamps.

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py [hunk 4]
```diff
@@ -3698,17 +3705,24 @@ def get_modified_remote_data_command(client: Client, params: dict[str, str],
...
-    filter_closed = f'id <= {highest_fetched_id} AND status=closed AND close_time > {last_update}'
+    last_update_modified = ctx.get(LAST_MIRROR_KEY, 0)
+    if not last_update_modified:
+        # This is the first mirror. We get the last update of the latest incident with a window of 5 minutes
+        last_update = dateparser.parse(remote_args.last_update)
+        if not last_update:
+            last_update = datetime.now()
+        last_update -= timedelta(minutes=5)
+        last_update_modified = int(last_update.timestamp() * 1000)
+    last_update_closed = ctx.get(LAST_MIRROR_CLOSED_KEY, last_update_modified)
```
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:3721]: Maybe first check `if LAST_MIRROR_CLOSED_KEY in ctx`,
and only if it's not, then do the whole `last_update` and `last_update_modified` calculation(s)?

**Review by tkatzir — APPROVED** [review]
Very good!
Approved, with a comment.

**Pull request merged**